go/pkg/casng/batching.go

```diff
@@ -329,5 +434,215 @@ func (u *BatchingUploader) DigestTree(ctx context.Context, root impath.Absolute,
 // All requests must share the same filter. Digest fields on the requests are ignored to ensure proper hierarchy caching via the internal digestion process.
 // remoteWorkingDir replaces workingDir inside the merkle tree such that the server is only aware of remoteWorkingDir.
 func (u *BatchingUploader) UploadTree(ctx context.Context, execRoot impath.Absolute, workingDir, remoteWorkingDir impath.Relative, reqs ...UploadRequest) (rootDigest digest.Digest, uploaded []digest.Digest, stats Stats, err error) {
-	panic("not yet implemented")
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload.tree; reqs=%d", len(reqs))
+	defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.tree.done")
+
+	if len(reqs) == 0 {
+		return
+	}
+
+	// 1st, Preprocess the set to deduplicate by ancestor and generate a deterministic filter ID.
+
+	// Multiple reqs sets may share some of the paths which would cause the the u.dirChildren lookup below to mix children from two different sets which would corrupt the merkle tree.
+	// Updating the filterID for the set to a deterministic one ensures it gets its unique keys that are still shared between identical sets.
+	// The deterministic ID is the digest of the sorted list of paths concatenated with the ID of the original filter.
+	filter := reqs[0].Exclude
+	filterID := filter.String()
+	paths := make([]string, 0, len(reqs)+1)
+	paths = append(paths, filterID)
+
+	// An ancestor directory takes precedence over all of its descendants to ensure unlisted files are also included in traversal.
+	// Sorting the requests by path length ensures ancestors appear before descendants.
+	sort.Slice(reqs, func(i, j int) bool { return len(reqs[i].Path.String()) < len(reqs[j].Path.String()) })
```

Comment:
just checking: are the paths normalized? e.g. is there no chance that the request would contain `..` component in the path?

Reply:
Yes. `impath` package ensures every path it creates is clean and immutable.